File: Sources/Sentry/SentryANRTrackerV2.m
#import "SentryANRTrackerV2.h"

#if SENTRY_HAS_UIKIT

#    import "SentryANRStoppedResultInternal.h"
#    import "SentryANRTrackerInternalDelegate.h"
#    import "SentryLogC.h"
#    import "SentrySwift.h"
#    import "SentryTime.h"
#    import <stdatomic.h>

NS_ASSUME_NONNULL_BEGIN

typedef NS_ENUM(NSInteger, SentryANRTrackerState) {
    kSentryANRTrackerNotRunning = 1,
    kSentryANRTrackerRunning,
    kSentryANRTrackerStarting,
    kSentryANRTrackerStopping
};

@interface SentryANRTrackerV2 ()

@property (nonatomic, strong) id<SentryApplicationStateProvider> applicationStateProvider;
@property (nonatomic, strong) SentryDispatchQueueWrapper *dispatchQueueWrapper;
@property (nonatomic, strong) SentryThreadWrapper *threadWrapper;
@property (nonatomic, strong) NSHashTable<id<SentryANRTrackerInternalDelegate>> *listeners;
@property (nonatomic, strong) SentryFramesTracker *framesTracker;
@property (nonatomic, assign) NSTimeInterval timeoutInterval;

@end

@implementation SentryANRTrackerV2 {
    NSObject *threadLock;
    SentryANRTrackerState state;
}

/**
 * Returns the frames delay to trust for hang classification, and via @c trustedFramesCount the
 * matching frames count when passed non NULL.
 *
 * When the main thread drains its queue, the run loop turns, and an active display link fires
 * within roughly one frame duration afterwards. So when the main thread is responsive, an ongoing
 * frame gap larger than the @c sleepInterval means the app isn't rendering although the main
 * thread isn't blocked, and the ongoing frame's delay must be ignored. Recorded delayed frames
 * stay trusted because they actually rendered. The ongoing frame always contributes one frame to
 * @c framesContributingToDelayCount (see @c getFramesDelayObjC), so ignoring its delay also
 * removes it from the trusted frames count.
 */
static CFTimeInterval
getTrustedFramesDelayDuration(SentryFramesDelayResultSPI *framesDelay,
    BOOL isMainThreadUnresponsive, NSTimeInterval sleepInterval,
    NSUInteger *_Nullable trustedFramesCount)
{
    CFTimeInterval delayDuration = framesDelay.delayDuration;
    NSUInteger framesCount = framesDelay.framesContributingToDelayCount;

    if (!isMainThreadUnresponsive && framesDelay.ongoingFrameDelayDuration > sleepInterval) {
        delayDuration -= framesDelay.ongoingFrameDelayDuration;
        framesCount -= 1;
    }

    if (trustedFramesCount != NULL) {
        *trustedFramesCount = framesCount;
    }
    return delayDuration;
}

- (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
{
    return [self
         initWithTimeoutInterval:timeoutInterval
        applicationStateProvider:SentryDependencyContainer.sharedInstance.applicationStateProvider
            dispatchQueueWrapper:SentryDependencyContainer.sharedInstance.dispatchQueueWrapper
                   threadWrapper:SentryDependencyContainer.sharedInstance.threadWrapper
                   framesTracker:SentryDependencyContainer.sharedInstance.framesTracker];
}

- (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
               applicationStateProvider:(id<SentryApplicationStateProvider>)applicationStateProvider
                   dispatchQueueWrapper:(SentryDispatchQueueWrapper *)dispatchQueueWrapper
                          threadWrapper:(SentryThreadWrapper *)threadWrapper
                          framesTracker:(SentryFramesTracker *)framesTracker
{
    if (self = [super init]) {
        self.timeoutInterval = timeoutInterval;
        self.applicationStateProvider = applicationStateProvider;
        self.dispatchQueueWrapper = dispatchQueueWrapper;
        self.threadWrapper = threadWrapper;
        self.framesTracker = framesTracker;
        self.listeners = [NSHashTable weakObjectsHashTable];
        threadLock = [[NSObject alloc] init];
        state = kSentryANRTrackerNotRunning;
    }
    return self;
}

- (void)detectANRs
{
    NSUUID *threadID = [NSUUID UUID];

    @synchronized(threadLock) {
        [self.threadWrapper threadStarted:threadID];

        if (state != kSentryANRTrackerStarting) {
            [self.threadWrapper threadFinished:threadID];
            return;
        }

        NSThread.currentThread.name = @"io.sentry.app-hang-tracker";
        state = kSentryANRTrackerRunning;
    }

    id<SentryCurrentDateProvider> dateProvider
        = SentryDependencyContainer.sharedInstance.dateProvider;

    BOOL reported = NO;

    // Ticks since the main thread last drained the main queue. When the ticks reach the
    // reportThreshold the main thread is treated as unresponsive; see the heartbeat block below.
    __block atomic_int ticksSinceLastMainQueueDrain = 0;

    NSInteger reportThreshold = 5;
    NSTimeInterval sleepInterval = self.timeoutInterval / reportThreshold;
    uint64_t sleepIntervalInNanos = timeIntervalToNanoseconds(sleepInterval);
    uint64_t timeoutIntervalInNanos = timeIntervalToNanoseconds(self.timeoutInterval);

    uint64_t appHangStoppedInterval = timeIntervalToNanoseconds(sleepInterval * 2);
    CFTimeInterval appHangStoppedFrameDelayThreshold
        = nanosecondsToTimeInterval(appHangStoppedInterval) * 0.2;

    uint64_t lastAppHangStoppedSystemTime = dateProvider.systemTime - timeoutIntervalInNanos;
    uint64_t lastAppHangStartedSystemTime = 0;

    // Track time to exclude from the hang duration calculation. While the app is in the
    // background, the frames tracker can't provide frame delay data, e.g. because it's paused, or
    // the OS suspends the app, the app can't be hanging, but the system time keeps ticking.
    BOOL isExcludingTime = NO;
    uint64_t excludedTimeStartSystemTime = 0;
    uint64_t accumulatedExcludedTime = 0;

    // Canceling the thread can take up to sleepInterval.
    while (YES) {
        @synchronized(threadLock) {
            if (state != kSentryANRTrackerRunning) {
                break;
            }
        }

        NSDate *sleepDeadline = [[dateProvider date] dateByAddingTimeInterval:self.timeoutInterval];

        // Frame delay data alone can't distinguish a blocked main thread from an app that
        // legitimately stopped rendering while staying active, e.g. a CarPlay scene keeping the
        // app alive while the phone is locked, or the proximity sensor blanking the screen; see
        // https://github.com/getsentry/sentry-cocoa/issues/8317. This heartbeat provides the
        // missing signal: when the main thread drains its queue, it isn't blocked.
        atomic_fetch_add_explicit(&ticksSinceLastMainQueueDrain, 1, memory_order_relaxed);
        [self.dispatchQueueWrapper dispatchAsyncOnMainQueueIfNotMainThread:^{
            atomic_store_explicit(&ticksSinceLastMainQueueDrain, 0, memory_order_relaxed);
        }];

        [self.threadWrapper sleepForTimeInterval:sleepInterval];

        @synchronized(threadLock) {
            if (state != kSentryANRTrackerRunning) {
                break;
            }
        }

        BOOL isInForeground = [self.applicationStateProvider isApplicationInForeground];

        if (!isInForeground) {
            SENTRY_LOG_DEBUG(@"Ignoring potential app hangs because the app is in the background");

            // Start excluding time from the hang duration. Background time is one of three
            // exclusion causes; see the declaration of isExcludingTime.
            if (reported && !isExcludingTime) {
                isExcludingTime = YES;
                excludedTimeStartSystemTime = dateProvider.systemTime;
            }

            continue;
        }

        // The sleepDeadline should be roughly executed after the timeoutInterval even if there is
        // an AppHang. If the app gets suspended this thread could sleep and wake up again. To avoid
        // false positives, we don't report AppHangs if the delta is too big.
        NSTimeInterval deltaFromNowToSleepDeadline =
            [[dateProvider date] timeIntervalSinceDate:sleepDeadline];

        if (deltaFromNowToSleepDeadline >= self.timeoutInterval) {
            SENTRY_LOG_DEBUG(@"Ignoring App Hang because the delta is too big: %f.",
                deltaFromNowToSleepDeadline);

            // The thread slept way longer than expected, e.g. because the OS suspended the app.
            // Exclude the extra sleep time from the hang duration. When already excluding time,
            // the ongoing excluded time span covers the extra sleep time.
            if (reported && !isExcludingTime) {
                accumulatedExcludedTime += timeIntervalToNanoseconds(
                    deltaFromNowToSleepDeadline + self.timeoutInterval - sleepInterval);
            }
            continue;
        }

        uint64_t nowSystemTime = dateProvider.systemTime;

        int mainQueueDrainTicks
            = atomic_load_explicit(&ticksSinceLastMainQueueDrain, memory_order_relaxed);
        BOOL isMainThreadUnresponsive = mainQueueDrainTicks >= reportThreshold;

        if (reported) {

            uint64_t framesDelayStartSystemTime = nowSystemTime - appHangStoppedInterval;

            SentryFramesDelayResultSPI *framesDelay =
                [self.framesTracker getFramesDelaySPI:framesDelayStartSystemTime
                                   endSystemTimestamp:nowSystemTime];

            if (framesDelay.delayDuration == -1) {
                // Without frame delay data, e.g. because the frames tracker is paused after the
                // app resigned active, the app can't be hanging. Exclude that time from the hang
                // duration.
                if (!isExcludingTime) {
                    isExcludingTime = YES;
                    excludedTimeStartSystemTime = nowSystemTime;
                }
                continue;
            }

            if (isExcludingTime) {
                accumulatedExcludedTime += nowSystemTime - excludedTimeStartSystemTime;
                isExcludingTime = NO;
            }

            CFTimeInterval framesDelayDuration = getTrustedFramesDelayDuration(
                framesDelay, isMainThreadUnresponsive, sleepInterval, NULL);

            BOOL appHangStopped = framesDelayDuration < appHangStoppedFrameDelayThreshold;

            if (appHangStopped) {
                SENTRY_LOG_DEBUG(@"App hang stopped.");

                // As we check every sleepInterval if the app is hanging, the app could already be
                // hanging for almost the sleepInterval until we detect it and it could already
                // stopped hanging almost a sleepInterval until we again detect it's not.
                //
                // Subtract any time during which the app couldn't have been hanging, such as
                // time in background, without frame delay data, or while the OS suspended the
                // app. The system time continues to tick during that time, but we don't want to
                // include it in the reported duration.
                uint64_t elapsedSystemTime = nowSystemTime - lastAppHangStartedSystemTime;
                uint64_t observedElapsedTime = elapsedSystemTime > accumulatedExcludedTime
                    ? elapsedSystemTime - accumulatedExcludedTime
                    : 0;
                uint64_t appHangDurationNanos = timeoutIntervalInNanos + observedElapsedTime;

                NSTimeInterval appHangDurationMinimum
                    = nanosecondsToTimeInterval(appHangDurationNanos - sleepIntervalInNanos);
                NSTimeInterval appHangDurationMaximum
                    = nanosecondsToTimeInterval(appHangDurationNanos + sleepIntervalInNanos);

                lastAppHangStoppedSystemTime = nowSystemTime;
                reported = NO;
                isExcludingTime = NO;
                accumulatedExcludedTime = 0;

                // The App Hang stopped, don't block the App Hangs thread or the main thread with
                // calling ANRStopped listeners.
                [self.dispatchQueueWrapper dispatchAsyncWithBlock:^{
                    [self ANRStopped:appHangDurationMinimum to:appHangDurationMaximum];
                }];
            }

            continue;
        }

        uint64_t lastAppHangLongEnoughInPastThreshold
            = lastAppHangStoppedSystemTime + timeoutIntervalInNanos;

        if (dateProvider.systemTime < lastAppHangLongEnoughInPastThreshold) {
            SENTRY_LOG_DEBUG(@"Ignoring app hang cause one happened recently.");
            continue;
        }

        uint64_t frameDelayStartSystemTime = nowSystemTime - timeoutIntervalInNanos;

        SentryFramesDelayResultSPI *framesDelayForTimeInterval =
            [self.framesTracker getFramesDelaySPI:frameDelayStartSystemTime
                               endSystemTimestamp:nowSystemTime];

        if (framesDelayForTimeInterval.delayDuration == -1) {
            continue;
        }

        NSUInteger framesContributingToDelayCount;
        CFTimeInterval delayDuration = getTrustedFramesDelayDuration(framesDelayForTimeInterval,
            isMainThreadUnresponsive, sleepInterval, &framesContributingToDelayCount);

        uint64_t framesDelayForTimeIntervalInNanos = timeIntervalToNanoseconds(delayDuration);

        BOOL isFullyBlocking = framesContributingToDelayCount == 1;

        if (isFullyBlocking && framesDelayForTimeIntervalInNanos >= timeoutIntervalInNanos) {
            SENTRY_LOG_WARN(@"App Hang detected: fully-blocking.");

            reported = YES;
            lastAppHangStartedSystemTime = dateProvider.systemTime;
            [self ANRDetected:kSentryANRTypeFullyBlocking];
        }

        NSTimeInterval nonFullyBlockingFramesDelayThreshold = self.timeoutInterval * 0.99;
        if (!isFullyBlocking && delayDuration > nonFullyBlockingFramesDelayThreshold) {

            SENTRY_LOG_WARN(@"App Hang detected: non-fully-blocking.");

            reported = YES;
            lastAppHangStartedSystemTime = dateProvider.systemTime;
            [self ANRDetected:kSentryANRTypeNonFullyBlocking];
        }
    }

    @synchronized(threadLock) {
        state = kSentryANRTrackerNotRunning;
        [self.threadWrapper threadFinished:threadID];
    }
}

- (void)ANRDetected:(SentryANRTypeInternal)type
{
    NSArray *localListeners;
    @synchronized(self.listeners) {
        localListeners = [self.listeners allObjects];
    }

    for (id<SentryANRTrackerInternalDelegate> target in localListeners) {
        [target anrDetected:type];
    }
}

- (void)ANRStopped:(NSTimeInterval)hangDurationMinimum to:(NSTimeInterval)hangDurationMaximum
{
    NSArray *targets;
    @synchronized(self.listeners) {
        targets = [self.listeners allObjects];
    }

    SentryANRStoppedResultInternal *result =
        [[SentryANRStoppedResultInternal alloc] initWithMinDuration:hangDurationMinimum
                                                        maxDuration:hangDurationMaximum];
    for (id<SentryANRTrackerInternalDelegate> target in targets) {
        [target anrStopped:result];
    }
}

- (void)addListener:(id<SentryANRTrackerInternalDelegate>)listener
{
    @synchronized(self.listeners) {
        [self.listeners addObject:listener];

        @synchronized(threadLock) {
            if (self.listeners.count > 0 && state == kSentryANRTrackerNotRunning) {
                if (state == kSentryANRTrackerNotRunning) {
                    state = kSentryANRTrackerStarting;
                    [NSThread detachNewThreadSelector:@selector(detectANRs)
                                             toTarget:self
                                           withObject:nil];
                }
            }
        }
    }
}

- (void)removeListener:(id<SentryANRTrackerInternalDelegate>)listener
{
    @synchronized(self.listeners) {
        [self.listeners removeObject:listener];

        if (self.listeners.count == 0) {
            [self stop];
        }
    }
}

- (void)clear
{
    @synchronized(self.listeners) {
        [self.listeners removeAllObjects];
        [self stop];
    }
}

- (void)stop
{
    @synchronized(threadLock) {
        SENTRY_LOG_INFO(@"Stopping App Hang detection");
        state = kSentryANRTrackerStopping;
    }
}

@end

NS_ASSUME_NONNULL_END

#endif // SENTRY_HAS_UIKIT

File: Sources/Swift/Integrations/AppHangTracking/SentryANRTrackerV2.swift
#if !SDK_V10 && (os(iOS) || os(tvOS) || os(visionOS)) && !SENTRY_NO_UI_FRAMEWORK
internal import _SentryPrivate
import Foundation

/// Detects fully and non-fully blocking app hangs from frame delays on a dedicated watchdog thread.
final class SentryANRTrackerV2: SentryANRTrackerInternalProtocol {
    private enum Lifecycle {
        case notRunning, running, starting, stopping
    }

    private let applicationStateProvider: SentryApplicationStateProvider
    private let dispatchQueueWrapper: SentryDispatchQueueWrapper
    private let threadWrapper: SentryThreadWrapper
    private let listeners = SentryMutex(NSHashTable<AnyObject>.weakObjects())
    private let framesTracker: SentryFramesTracker
    private let timeoutInterval: TimeInterval
    private let threadState = SentryMutex<Lifecycle>(.notRunning)

    convenience init(timeoutInterval: TimeInterval) {
        let dependencies = SentryDependencyContainer.sharedInstance()
        self.init(timeoutInterval: timeoutInterval,
                  applicationStateProvider: dependencies.applicationStateProvider,
                  dispatchQueueWrapper: dependencies.dispatchQueueWrapper,
                  threadWrapper: dependencies.threadWrapper,
                  framesTracker: dependencies.framesTracker)
    }

    init(timeoutInterval: TimeInterval,
         applicationStateProvider: SentryApplicationStateProvider,
         dispatchQueueWrapper: SentryDispatchQueueWrapper,
         threadWrapper: SentryThreadWrapper,
         framesTracker: SentryFramesTracker) {
        self.timeoutInterval = timeoutInterval
        self.applicationStateProvider = applicationStateProvider
        self.dispatchQueueWrapper = dispatchQueueWrapper
        self.threadWrapper = threadWrapper
        self.framesTracker = framesTracker
    }

    // Keep the watchdog flow together to make the Objective-C conversion directly comparable.
    // swiftlint:disable:next function_body_length cyclomatic_complexity
    private func detectANRs() {
        let threadID = UUID()
        let shouldRun = threadState.withLock { state in
            threadWrapper.threadStarted(threadID)
            if state != .starting {
                threadWrapper.threadFinished(threadID)
                return false
            }
            Thread.current.name = "io.sentry.app-hang-tracker"
            state = .running
            return true
        }
        guard shouldRun else { return }

        let dateProvider = SentryDependencyContainer.sharedInstance().dateProvider
        var reported = false

        let reportThreshold = 5
        let sleepInterval = timeoutInterval / Double(reportThreshold)
        let sleepIntervalInNanos = timeIntervalToNanoseconds(sleepInterval)
        let timeoutIntervalInNanos = timeIntervalToNanoseconds(timeoutInterval)
        let appHangStoppedInterval = timeIntervalToNanoseconds(sleepInterval * 2)
        let appHangStoppedFrameDelayThreshold = nanosecondsToTimeInterval(appHangStoppedInterval) * 0.2

        // Preserve the unsigned timestamp arithmetic used by the Objective-C implementation.
        var lastAppHangStoppedSystemTime = dateProvider.systemTime() &- timeoutIntervalInNanos
        var lastAppHangStartedSystemTime: UInt64 = 0

        // Exclude background time from an ongoing hang's duration while system time keeps ticking.
        var wasInBackground = false
        var wentToBackgroundSystemTime: UInt64 = 0
        var accumulatedBackgroundTime: UInt64 = 0

        // Cancelling the thread can take up to sleepInterval.
        while true {
            if threadState.withLock({ $0 != .running }) {
                break
            }

            let sleepDeadline = dateProvider.date().addingTimeInterval(timeoutInterval)
            threadWrapper.sleep(forTimeInterval: sleepInterval)

            if threadState.withLock({ $0 != .running }) {
                break
            }

            let isInForeground = applicationStateProvider.isApplicationInForeground
            if !isInForeground {
                SentrySDKLog.debug("Ignoring potential app hangs because the app is in the background")
                if reported && !wasInBackground {
                    wasInBackground = true
                    wentToBackgroundSystemTime = dateProvider.systemTime()
                }
                continue
            }

            if reported && wasInBackground {
                let backgroundTime = dateProvider.systemTime() &- wentToBackgroundSystemTime
                accumulatedBackgroundTime &+= backgroundTime
                wasInBackground = false
            }

            // A suspended app can wake much later than expected. Do not report that as an app hang.
            let deltaFromNowToSleepDeadline = dateProvider.date().timeIntervalSince(sleepDeadline)
            if deltaFromNowToSleepDeadline >= timeoutInterval {
                SentrySDKLog.debug("Ignoring App Hang because the delta is too big: \(deltaFromNowToSleepDeadline).")
                continue
            }

            let nowSystemTime = dateProvider.systemTime()
            if reported {
                let framesDelayStartSystemTime = nowSystemTime &- appHangStoppedInterval
                let framesDelay = framesTracker.getFramesDelay(framesDelayStartSystemTime, endSystemTimestamp: nowSystemTime)
                if framesDelay.delayDuration == -1 {
                    continue
                }

                let appHangStopped = framesDelay.delayDuration < appHangStoppedFrameDelayThreshold
                if appHangStopped {
                    SentrySDKLog.debug("App hang stopped.")

                    // Polling can detect the beginning and end up to one sleep interval late.
                    // Subtract background time, during which system time continues to tick.
                    let elapsedSystemTime = nowSystemTime &- lastAppHangStartedSystemTime
                    let foregroundElapsedTime = elapsedSystemTime > accumulatedBackgroundTime
                        ? elapsedSystemTime - accumulatedBackgroundTime : 0
                    let appHangDurationNanos = timeoutIntervalInNanos &+ foregroundElapsedTime
                    let appHangDurationMinimum = nanosecondsToTimeInterval(appHangDurationNanos &- sleepIntervalInNanos)
                    let appHangDurationMaximum = nanosecondsToTimeInterval(appHangDurationNanos &+ sleepIntervalInNanos)

                    lastAppHangStoppedSystemTime = nowSystemTime
                    reported = false
                    wasInBackground = false
                    accumulatedBackgroundTime = 0

                    // Keep listener work off both the watchdog and main threads.
                    dispatchQueueWrapper.dispatchAsync { [self] in
                        anrStopped(appHangDurationMinimum, to: appHangDurationMaximum)
                    }
                }
                continue
            }

            let lastAppHangLongEnoughInPastThreshold = lastAppHangStoppedSystemTime &+ timeoutIntervalInNanos
            if dateProvider.systemTime() < lastAppHangLongEnoughInPastThreshold {
                SentrySDKLog.debug("Ignoring app hang cause one happened recently.")
                continue
            }

            let frameDelayStartSystemTime = nowSystemTime &- timeoutIntervalInNanos
            let framesDelayForTimeInterval = framesTracker.getFramesDelay(frameDelayStartSystemTime, endSystemTimestamp: nowSystemTime)
            if framesDelayForTimeInterval.delayDuration == -1 {
                continue
            }

            let framesDelayForTimeIntervalInNanos = timeIntervalToNanoseconds(framesDelayForTimeInterval.delayDuration)
            let isFullyBlocking = framesDelayForTimeInterval.framesContributingToDelayCount == 1
            if isFullyBlocking && framesDelayForTimeIntervalInNanos >= timeoutIntervalInNanos {
                SentrySDKLog.warning("App Hang detected: fully-blocking.")
                reported = true
                lastAppHangStartedSystemTime = dateProvider.systemTime()
                anrDetected(.fullyBlocking)
            }

            let nonFullyBlockingFramesDelayThreshold = timeoutInterval * 0.99
            if !isFullyBlocking && framesDelayForTimeInterval.delayDuration > nonFullyBlockingFramesDelayThreshold {
                SentrySDKLog.warning("App Hang detected: non-fully-blocking.")
                reported = true
                lastAppHangStartedSystemTime = dateProvider.systemTime()
                anrDetected(.nonFullyBlocking)
            }
        }

        threadState.withLock { state in
            state = .notRunning
            threadWrapper.threadFinished(threadID)
        }
    }

    private func anrDetected(_ type: SentryANRType) {
        let localListeners = listeners.withLock {
            $0.allObjects.compactMap { $0 as? SentryANRTrackerInternalDelegate }
        }
        for target in localListeners {
            target.anrDetected(type)
        }
    }

    private func anrStopped(_ hangDurationMinimum: TimeInterval, to hangDurationMaximum: TimeInterval) {
        let targets = listeners.withLock {
            $0.allObjects.compactMap { $0 as? SentryANRTrackerInternalDelegate }
        }
        let result = SentryANRStoppedResultInternal(minDuration: hangDurationMinimum, maxDuration: hangDurationMaximum)
        for target in targets {
            target.anrStopped(result)
        }
    }

    func addListener(_ listener: SentryANRTrackerInternalDelegate) {
        listeners.withLock { listeners in
            listeners.add(listener)
            threadState.withLock { state in
                if listeners.count > 0 && state == .notRunning {
                    state = .starting
                    // NSThread retained its target. Keep that lifetime and start under the same locks.
                    Thread.detachNewThread { [self] in
                        detectANRs()
                    }
                }
            }
        }
    }

    func removeListener(_ listener: SentryANRTrackerInternalDelegate) {
        listeners.withLock { listeners in
            listeners.remove(listener)
            if listeners.count == 0 {
                stop()
            }
        }
    }

    func clear() {
        listeners.withLock { listeners in
            listeners.removeAllObjects()
            stop()
        }
    }

    private func stop() {
        threadState.withLock { state in
            SentrySDKLog.info("Stopping App Hang detection")
            state = .stopping
        }
    }
}
#endif // !SDK_V10 && UIKit
